// @progress/kendo-angular-inputs/dist/es2015/main.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
export { SliderComponent } from './slider/slider.component';
export { RangeSliderComponent } from './rangeslider/rangeslider.component';
export { LabelTemplateDirective } from './sliders-common/label-template.directive';
export { SwitchComponent } from './switch/switch.component';
export { TextBoxContainerComponent } from './textbox/textbox-container.component';
export { TextBoxDirective } from './textbox/textbox.directive';
export { TextAreaDirective } from './shared/textarea.directive';
export { NumericTextBoxComponent } from './numerictextbox/numerictextbox.component';
export { NumericTextBoxCustomMessagesComponent } from './numerictextbox/localization/custom-messages.component';
export { MaskedTextBoxComponent } from './maskedtextbox/maskedtextbox.component';
export { InputsModule } from './inputs.module';
export { SliderTicksComponent } from './sliders-common/slider-ticks.component';
export { SliderModule } from './slider.module';
export { RangeSliderModule } from './rangeslider.module';
export { SwitchModule } from './switch.module';
export { NumericTextBoxModule } from './numerictextbox.module';
export { MaskedTextBoxModule } from './maskedtextbox.module';
export { TextBoxModule } from './textbox.module';
export { TextAreaModule } from './textarea.module';
export { ColorPickerComponent } from './colorpicker/colorpicker.component';
export { ColorPaletteComponent } from './colorpicker/color-palette.component';
export { ColorGradientComponent } from './colorpicker/color-gradient.component';
export { ColorPickerModule } from './colorpicker.module';
export { ActiveColorClickEvent } from './colorpicker/models';
export { CheckBoxDirective } from './checkbox/checkbox.directive';
export { RadioButtonDirective } from './radiobutton/radiobutton.directive';
export { HintComponent } from './formfield/hint.component';
export { ErrorComponent } from './formfield/error.component';
export { FormFieldComponent } from './formfield/formfield.component';
export { FormFieldModule } from './formfield.module';
//TextBox Component
export { TextBoxComponent } from './textbox/textbox.component';
export { TextBoxPrefixTemplateDirective } from './textbox/textbox-prefix.directive';
export { TextBoxSuffixTemplateDirective } from './textbox/textbox-suffix.directive';
//TextArea Component
export { TextAreaComponent } from './textarea/textarea.component';
export { TextAreaSuffixComponent } from './textarea/textarea-suffix.component';
//Shared
export { InputSeparatorComponent } from './shared/input-separator.component';
export { SharedModule } from './shared.module';
export { LocalizedNumericTextBoxMessagesDirective } from './numerictextbox/localization/localized-numerictextbox-messages.directive';
export { LocalizedSliderMessagesDirective } from './slider/localization/localized-slider-messages.directive';
export { LocalizedRangeSliderMessagesDirective } from './rangeslider/localization/localized-rangeslider-messages.directive';
export { LocalizedSwitchMessagesDirective } from './switch/localization/localized-switch-messages.directive';
