// @igniteui/angular-schematics/src/app-projects/index_spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
const testing_1 = require("@angular-devkit/schematics/testing");
const path = require("path");
const collectionPath = path.join(__dirname, "../collection.json");
describe("app-projects", () => {
    it("works", () => {
        const runner = new testing_1.SchematicTestRunner("schematics", collectionPath);
        const tree = schematics_1.Tree.empty();
        const mockOptions = {
            name: "mock-name",
            projTemplate: {
                generateConfig: jasmine.createSpy("generateConfig").and.returnValue({}),
                id: "mock-id",
                templatePaths: ["mock-template"]
            },
            theme: "mock-theme"
        };
        spyOn(tree, "read").and.returnValue(`Mock package content "igniteui-cli":`);
        spyOn(tree, "overwrite");
        runner.runSchematic("app-projects", mockOptions, tree);
        expect(mockOptions.projTemplate.generateConfig).toHaveBeenCalled();
        expect(tree.read).toHaveBeenCalledWith("./package.json");
        expect(tree.overwrite).toHaveBeenCalledWith("./package.json", `Mock package content "@igniteui/angular-schematics":`);
    });
});
