// @progress/kendo-angular-gauges/dist/npm/radial-gauge/radial-gauge.directives.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var radial_gauge_component_1 = require("./radial-gauge.component");
exports.RadialGaugeComponent = radial_gauge_component_1.RadialGaugeComponent;
var gauge_area_component_1 = require("./gauge-area.component");
exports.RadialGaugeAreaComponent = gauge_area_component_1.RadialGaugeAreaComponent;
var scale_component_1 = require("./scale.component");
exports.RadialScaleComponent = scale_component_1.RadialScaleComponent;
var labels_component_1 = require("./labels.component");
exports.RadialLabelsComponent = labels_component_1.RadialLabelsComponent;
var pointers_component_1 = require("./pointers.component");
exports.RadialPointersComponent = pointers_component_1.RadialPointersComponent;
var pointer_component_1 = require("./pointer.component");
exports.RadialPointerComponent = pointer_component_1.RadialPointerComponent;
var range_component_1 = require("./range.component");
exports.RadialRangeComponent = range_component_1.RadialRangeComponent;
var ranges_component_1 = require("./ranges.component");
exports.RadialRangesComponent = ranges_component_1.RadialRangesComponent;
