// @progress/kendo-angular-gauges/dist/npm/arc-gauge/arc-gauge.directives.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var arc_gauge_component_1 = require("./arc-gauge.component");
exports.ArcGaugeComponent = arc_gauge_component_1.ArcGaugeComponent;
var gauge_area_component_1 = require("./gauge-area.component");
exports.ArcGaugeAreaComponent = gauge_area_component_1.ArcGaugeAreaComponent;
var scale_component_1 = require("./scale.component");
exports.ArcScaleComponent = scale_component_1.ArcScaleComponent;
var labels_component_1 = require("./labels.component");
exports.ArcLabelsComponent = labels_component_1.ArcLabelsComponent;
var arc_center_template_directive_1 = require("./arc-center-template.directive");
exports.ArcCenterTemplateDirective = arc_center_template_directive_1.ArcCenterTemplateDirective;
var colors_component_1 = require("./colors.component");
exports.ColorsComponent = colors_component_1.ColorsComponent;
var color_component_1 = require("./color.component");
exports.ColorComponent = color_component_1.ColorComponent;
