// @progress/kendo-angular-gauges/dist/npm/arc-gauge/arc-gauge.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var services_1 = require("../services");
var kendo_angular_intl_1 = require("@progress/kendo-angular-intl");
var kendo_angular_l10n_1 = require("@progress/kendo-angular-l10n");
var base_components_1 = require("../base-components");
var arc_center_template_directive_1 = require("./arc-center-template.directive");
var kendo_charts_1 = require("@progress/kendo-charts");
var kendo_drawing_1 = require("@progress/kendo-drawing");
/* tslint:disable:use-life-cycle-interface */
/**
 * Represents the [Kendo UI ArcGauge component for Angular]({% slug overview_arcgauge_gauges %}).
 *
 * @example
 * ```ts
 * import { Component } from '@angular/core';
 *
 * _@Component({
 *   selector: 'my-app',
 *   template: `
 *      <kendo-arcgauge [value]="value" [scale]="{ max: 100 }">
 *          <ng-template kendoArcGaugeCenterTemplate let-value="value">
 *              {{ value }}%
 *          </ng-template>
 *      </kendo-arcgauge>
 *   `
 * })
 * class AppComponent {
 *   public value: number = 10;
 * }
 *
 * ```
 */
var ArcGaugeComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ArcGaugeComponent, _super);
    function ArcGaugeComponent(changeDetector, configurationService, themeService, intlService, localizationService, element, renderer, ngZone) {
        var _this = _super.call(this, configurationService, themeService, intlService, localizationService, element, renderer, ngZone) || this;
        _this.changeDetector = changeDetector;
        _this.className = true;
        _this.centerTemplateContext = {};
        return _this;
    }
    ArcGaugeComponent.prototype.ngOnInit = function () {
        _super.prototype.ngOnInit.call(this);
        if (this.element) {
            this.renderer.setStyle(this.element.nativeElement, 'position', 'relative');
        }
    };
    ArcGaugeComponent.prototype.ngAfterViewChecked = function () {
        _super.prototype.ngAfterViewChecked.call(this);
        if (this.labelElement && !this.centerTemplate) {
            this.changeDetector.detectChanges();
        }
        else if (!this.labelElement && this.centerTemplate) {
            this.updateCenterTemplate();
        }
    };
    /**
     * Exports the ArcGauge as a Drawing `Scene`.
     *
     * @returns {Promise<Group>} - A promise that will be resolved with the export visual.
     */
    ArcGaugeComponent.prototype.exportVisual = function () {
        return kendo_drawing_1.drawDOM(this.element.nativeElement);
    };
    /**
     * Detects the size of the container and redraws the Gauge.
     * Resizing is automatic unless you set the `resizeRateLimit` option to `0`.
     */
    ArcGaugeComponent.prototype.resize = function () {
        _super.prototype.resize.call(this);
        this.updateCenterTemplate();
    };
    ArcGaugeComponent.prototype.createInstance = function (element, options, theme, context) {
        this.instance = new kendo_charts_1.ArcGauge(element, options, theme, context);
        this.updateElements();
    };
    ArcGaugeComponent.prototype.updateOptions = function () {
        _super.prototype.updateOptions.call(this);
        this.updateElements();
    };
    ArcGaugeComponent.prototype.setValues = function () {
        var value = this.configurationService.readValue();
        this.instance.value(value);
        this.updateCenterTemplate();
    };
    ArcGaugeComponent.prototype.updateElements = function () {
        this.resizeSensor.acceptSize();
        this.updateCenterTemplate();
    };
    ArcGaugeComponent.prototype.updateCenterTemplate = function () {
        if (!this.instance || !this.centerTemplate) {
            return;
        }
        this.centerTemplateContext.value = this.instance.value();
        this.centerTemplateContext.color = this.instance.currentColor();
        this.changeDetector.detectChanges();
        this.positionLabel();
    };
    ArcGaugeComponent.prototype.positionLabel = function () {
        if (!this.labelElement) {
            return;
        }
        var element = this.labelElement.nativeElement;
        var width = element.offsetWidth;
        var height = element.offsetHeight;
        var position = this.instance.centerLabelPosition(width, height);
        element.style.top = position.top + "px";
        element.style.left = position.left + "px";
    };
    //tslint:disable-next-line: no-empty
    ArcGaugeComponent.prototype.updateSize = function () {
    };
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Number)
    ], ArcGaugeComponent.prototype, "value", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String)
    ], ArcGaugeComponent.prototype, "color", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Array)
    ], ArcGaugeComponent.prototype, "colors", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Number)
    ], ArcGaugeComponent.prototype, "opacity", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], ArcGaugeComponent.prototype, "scale", void 0);
    tslib_1.__decorate([
        core_1.ContentChild(arc_center_template_directive_1.ArcCenterTemplateDirective, { static: false }),
        tslib_1.__metadata("design:type", arc_center_template_directive_1.ArcCenterTemplateDirective)
    ], ArcGaugeComponent.prototype, "centerTemplate", void 0);
    tslib_1.__decorate([
        core_1.ViewChild("label", { static: false }),
        tslib_1.__metadata("design:type", core_1.ElementRef)
    ], ArcGaugeComponent.prototype, "labelElement", void 0);
    tslib_1.__decorate([
        core_1.HostBinding('class.k-arcgauge'),
        tslib_1.__metadata("design:type", Boolean)
    ], ArcGaugeComponent.prototype, "className", void 0);
    ArcGaugeComponent = tslib_1.__decorate([
        core_1.Component({
            changeDetection: core_1.ChangeDetectionStrategy.OnPush,
            exportAs: 'kendoArcGauge',
            providers: [
                services_1.ConfigurationService,
                kendo_angular_l10n_1.LocalizationService,
                {
                    provide: kendo_angular_l10n_1.L10N_PREFIX,
                    useValue: 'kendo.arcgauge'
                }
            ],
            selector: 'kendo-arcgauge',
            template: "\n    <div #surface class='k-chart-surface'></div>\n    <div class=\"k-arcgauge-label\" *ngIf=\"centerTemplate\" #label>\n        <ng-template [ngTemplateOutlet]=\"centerTemplate.templateRef\" [ngTemplateOutletContext]=\"centerTemplateContext\"></ng-template>\n    </div>\n    <kendo-resize-sensor (resize)=\"onResize($event)\" [rateLimit]=\"resizeRateLimit\"></kendo-resize-sensor>\n  "
        }),
        tslib_1.__metadata("design:paramtypes", [core_1.ChangeDetectorRef,
            services_1.ConfigurationService,
            services_1.ThemeService,
            kendo_angular_intl_1.IntlService,
            kendo_angular_l10n_1.LocalizationService,
            core_1.ElementRef,
            core_1.Renderer2,
            core_1.NgZone])
    ], ArcGaugeComponent);
    return ArcGaugeComponent;
}(base_components_1.GaugeComponent));
exports.ArcGaugeComponent = ArcGaugeComponent;
