// @igniteui/angular-schematics/src/utils/theme-import.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
const angular_templates_1 = require("@igniteui/angular-templates");
const config_1 = require("@schematics/angular/utility/config");
const path = require("path");
function importDefaultTheme(tree) {
    const sourceRoot = getDefaultProject(tree).sourceRoot;
    if (tree.exists(path.join(sourceRoot, "styles.sass"))) {
        importDefaultThemeSass(tree, "sass");
        return tree;
    }
    else if (tree.exists(path.join(sourceRoot, "styles.scss"))) {
        importDefaultThemeSass(tree, "scss");
        return tree;
    }
    importIgDefaultTheme(tree);
    return tree;
}
exports.importDefaultTheme = importDefaultTheme;
function addFontsToIndexHtml(tree) {
    const titillium = '<link href="https://fonts.googleapis.com/css?family=Titillium+Web" rel="stylesheet">';
    const materialIcons = '<link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">';
    const targetFile = getDefaultProjectBuildOptions(tree)["index"];
    if (tree.exists(targetFile)) {
        let content = tree.read(targetFile).toString();
        if (!content.includes(titillium)) {
            content = content.replace("</head>", `  ${titillium}\n</head>`);
        }
        if (!content.includes(materialIcons)) {
            content = content.replace("</head>", `  ${materialIcons}\n</head>`);
        }
        tree.overwrite(targetFile, content);
    }
    return tree;
}
exports.addFontsToIndexHtml = addFontsToIndexHtml;
function importDefaultThemeSass(tree, ext) {
    const igxPackage = angular_templates_1.resolveIgxPackage(angular_templates_1.NPM_PACKAGE);
    const sassImports = `
@import "~${igxPackage}/lib/core/styles/themes/index";
// Uncomment the following lines if you want to add a custom palette:
// $primary: #731963 !default;
// $secondary: #ce5712 !default;
// $app-palette: igx-palette($primary, $secondary);

/* autoprefixer grid: on */

@include igx-core();
@include igx-theme($default-palette);
`;
    const sourceRoot = getDefaultProject(tree).sourceRoot;
    const targetFile = `/${sourceRoot}/styles.${ext}`;
    let content = tree.read(targetFile).toString();
    if (!content.includes(sassImports)) {
        content += sassImports;
    }
    tree.overwrite(targetFile, content);
    return tree;
}
function importIgDefaultTheme(tree) {
    const targetFile = "/angular.json";
    const workspace = config_1.getWorkspace(tree);
    const importedStylesToBuild = importDefaultThemeToAngularWorkspace(workspace, "build");
    const importedStylesToTest = importDefaultThemeToAngularWorkspace(workspace, "test");
    if (importedStylesToBuild || importedStylesToTest) {
        tree.overwrite(targetFile, JSON.stringify(workspace, null, 2) + "\n");
    }
    return tree;
}
function getTargetedProjectOptions(project, target) {
    if (project.targets &&
        project.targets[target] &&
        project.targets[target].options) {
        return project.targets[target].options;
    }
    if (project.architect &&
        project.architect[target] &&
        project.architect[target].options) {
        return project.architect[target].options;
    }
    throw new schematics_1.SchematicsException(`Cannot determine the project's configuration for: ${target}`);
}
function getDefaultProject(tree) {
    const workspace = config_1.getWorkspace(tree);
    return workspace.projects[workspace.defaultProject];
}
exports.getDefaultProject = getDefaultProject;
function getDefaultProjectBuildOptions(tree) {
    return getTargetedProjectOptions(getDefaultProject(tree), "build");
}
exports.getDefaultProjectBuildOptions = getDefaultProjectBuildOptions;
function importDefaultThemeToAngularWorkspace(workspace, key) {
    const igxPackage = angular_templates_1.resolveIgxPackage(angular_templates_1.NPM_PACKAGE);
    const cssImport = `node_modules/${igxPackage}/styles/igniteui-angular.css`;
    const projectName = workspace.defaultProject;
    if (projectName) {
        if (!workspace.projects[projectName].architect) {
            workspace.projects[projectName].architect = {};
        }
        if (!workspace.projects[projectName].architect[key]) {
            workspace.projects[projectName].architect[key] = {};
        }
        if (!workspace.projects[projectName].architect[key].options) {
            workspace.projects[projectName].architect[key].options = {};
        }
        if (!workspace.projects[projectName].architect[key].options.styles) {
            workspace.projects[projectName].architect[key].options.styles = [];
        }
        if (!workspace.projects[projectName].architect[key].options.styles.includes(cssImport)) {
            workspace.projects[projectName].architect[key].options.styles.push(cssImport);
            return true;
        }
        return false;
    }
}
