// @progress/kendo-angular-inputs/dist/npm/formfield/hint.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var serial = 0;
/**
 * Represents a hint message that will be shown underneath a form-bound component.
 */
var HintComponent = /** @class */ (function () {
    function HintComponent() {
        /**
         * Specifies the alignment of the Hint message.
         *
         * The possible values are:
         * * (Default) `start`
         * * `end`
         */
        this.align = 'start';
        /**
         * @hidden
         */
        this.id = "kendo-hint-" + serial++;
        this.hostClass = true;
    }
    Object.defineProperty(HintComponent.prototype, "startClass", {
        get: function () {
            return this.align === 'start';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HintComponent.prototype, "endClass", {
        get: function () {
            return this.align === 'end';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HintComponent.prototype, "idAttribute", {
        get: function () {
            return this.id;
        },
        enumerable: true,
        configurable: true
    });
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String)
    ], HintComponent.prototype, "align", void 0);
    tslib_1.__decorate([
        core_1.HostBinding('class.k-form-hint'),
        tslib_1.__metadata("design:type", Boolean)
    ], HintComponent.prototype, "hostClass", void 0);
    tslib_1.__decorate([
        core_1.HostBinding('class.k-text-start'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [])
    ], HintComponent.prototype, "startClass", null);
    tslib_1.__decorate([
        core_1.HostBinding('class.k-text-end'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [])
    ], HintComponent.prototype, "endClass", null);
    tslib_1.__decorate([
        core_1.HostBinding('attr.id'),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [])
    ], HintComponent.prototype, "idAttribute", null);
    HintComponent = tslib_1.__decorate([
        core_1.Component({
            selector: 'kendo-formhint',
            template: "\n        <ng-content></ng-content>\n    "
        })
    ], HintComponent);
    return HintComponent;
}());
exports.HintComponent = HintComponent;
