// @progress/kendo-angular-gauges/dist/npm/base-components.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var collection_component_1 = require("./base-components/collection.component");
exports.CollectionComponent = collection_component_1.CollectionComponent;
var collection_item_component_1 = require("./base-components/collection-item.component");
exports.CollectionItemComponent = collection_item_component_1.CollectionItemComponent;
var gauge_component_1 = require("./base-components/gauge.component");
exports.GaugeComponent = gauge_component_1.GaugeComponent;
var gauge_area_component_1 = require("./base-components/gauge-area.component");
exports.GaugeAreaComponent = gauge_area_component_1.GaugeAreaComponent;
var labels_component_1 = require("./base-components/labels.component");
exports.LabelsComponent = labels_component_1.LabelsComponent;
var range_component_1 = require("./base-components/range.component");
exports.RangeComponent = range_component_1.RangeComponent;
var scale_component_1 = require("./base-components/scale.component");
exports.ScaleComponent = scale_component_1.ScaleComponent;
var settings_component_1 = require("./base-components/settings.component");
exports.SettingsComponent = settings_component_1.SettingsComponent;
