// @progress/kendo-angular-gauges/dist/es/services/configuration.service.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
function isObject(value) {
    return typeof value === "object";
}
function diffOptions(original, current) {
    if (Object.keys(original).length !== Object.keys(current).length) {
        return true;
    }
    for (var field in original) {
        if (field !== 'value' && original.hasOwnProperty(field)) {
            var originalValue = original[field];
            var currentValue = current[field];
            var diff = isObject(originalValue) && isObject(currentValue) ?
                diffOptions(originalValue, currentValue) : originalValue !== currentValue;
            if (diff) {
                return true;
            }
        }
    }
}
function diffPointerOptions(original, current) {
    if (!original || !current) {
        return true;
    }
    original = [].concat(original);
    current = [].concat(current);
    if (original.length !== current.length) {
        return true;
    }
    for (var idx = 0; idx < original.length; idx++) {
        if (diffOptions(original[idx], current[idx])) {
            return true;
        }
    }
}
/**
 * @hidden
 */
var ConfigurationService = /** @class */ (function () {
    function ConfigurationService() {
        this.options = {};
    }
    ConfigurationService.prototype.copyChanges = function (prefix, changes) {
        for (var propertyName in changes) {
            if (!changes.hasOwnProperty(propertyName)) {
                continue;
            }
            var value = changes[propertyName].currentValue;
            var optionName = (prefix ? prefix + '.' : '') + propertyName;
            this.set(optionName, value);
        }
    };
    ConfigurationService.prototype.read = function () {
        this.hasChanges = false;
        this.valueChange = false;
        return this.options;
    };
    ConfigurationService.prototype.readValues = function () {
        this.valueChange = false;
        var pointers = [].concat(this.options.pointer);
        return pointers.map(function (pointer) { return pointer.value; });
    };
    ConfigurationService.prototype.readValue = function () {
        this.valueChange = false;
        return this.options.value;
    };
    ConfigurationService.prototype.set = function (field, value) {
        var _a = this.optionContext(field), key = _a.key, options = _a.options;
        if (!this.hasChanges && (key === 'value' || (key === 'pointer' && !diffPointerOptions(this.options.pointer, value)))) {
            this.valueChange = true;
        }
        else {
            this.hasChanges = true;
            this.valueChange = false;
        }
        options[key] = value;
    };
    ConfigurationService.prototype.optionContext = function (field) {
        var parts = field.split('.');
        var options = this.options;
        var key = parts.shift();
        while (parts.length > 0) {
            options = options[key] = options[key] || {};
            key = parts.shift();
        }
        return { key: key, options: options };
    };
    ConfigurationService = tslib_1.__decorate([
        Injectable()
    ], ConfigurationService);
    return ConfigurationService;
}());
export { ConfigurationService };
