// @progress/kendo-angular-gauges/dist/es/services/theme.service.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
var template = "\n    <div class=\"k-var--gauge-pointer\"></div>\n    <div class=\"k-var--gauge-track\"></div>\n    <div class=\"k-var--normal-text-color\"></div>\n";
/**
 * @hidden
 */
var ThemeService = /** @class */ (function () {
    function ThemeService() {
    }
    ThemeService.prototype.read = function () {
        if (!this.options) {
            this.load();
        }
        return this.options;
    };
    ThemeService.prototype.load = function () {
        if (typeof document === 'undefined') {
            this.options = {};
            return;
        }
        var container = document.createElement('div');
        container.style.display = 'none';
        container.innerHTML = template;
        document.body.appendChild(container);
        try {
            var pointerColor = this.getColor(container, 'gauge-pointer');
            var rangePlaceholder = this.getColor(container, 'gauge-track');
            var textColor = this.getColor(container, 'normal-text-color');
            this.options = {
                pointer: {
                    color: pointerColor
                },
                scale: {
                    labels: {
                        color: textColor
                    },
                    rangePlaceholderColor: rangePlaceholder,
                    minorTicks: {
                        color: textColor
                    },
                    majorTicks: {
                        color: textColor
                    },
                    line: {
                        color: textColor
                    }
                }
            };
        }
        finally {
            document.body.removeChild(container);
        }
    };
    ThemeService.prototype.getColor = function (container, varName) {
        return window.getComputedStyle(container.querySelector(".k-var--" + varName)).backgroundColor;
    };
    ThemeService = tslib_1.__decorate([
        Injectable()
    ], ThemeService);
    return ThemeService;
}());
export { ThemeService };
